sys/arm64/vmm/vmm.c

```diff
@@ -0,0 +1,1818 @@
+
+struct vm *
+vcpu_vm(struct vcpu *vcpu)
+{
+	return (vcpu->vm);
+}
+
+int
+vcpu_vcpuid(struct vcpu *vcpu)
+{
+	return (vcpu->vcpuid);
+}
+
+void *
+vcpu_get_cookie(struct vcpu *vcpu)
+{
+	return (vcpu->cookie);
+}
+
+struct vcpu *
+vm_vcpu(struct vm *vm, int vcpuid)
+{
+	return (vm->vcpu[vcpuid]);
+}
+
+int
+vcpu_set_state(struct vcpu *vcpu, enum vcpu_state newstate, bool from_idle)
+{
+	int error;
+
+	vcpu_lock(vcpu);
+	error = vcpu_set_state_locked(vcpu, newstate, from_idle);
+	vcpu_unlock(vcpu);
+
+	return (error);
+}
+
+enum vcpu_state
+vcpu_get_state(struct vcpu *vcpu, int *hostcpu)
+{
+	enum vcpu_state state;
+
+	vcpu_lock(vcpu);
+	state = vcpu->state;
+	if (hostcpu != NULL)
+		*hostcpu = vcpu->hostcpu;
+	vcpu_unlock(vcpu);
+
+	return (state);
+}
+
+static void *
+_vm_gpa_hold(struct vm *vm, vm_paddr_t gpa, size_t len, int reqprot,
+    void **cookie)
+{
+	int i, count, pageoff;
+	struct mem_map *mm;
+	vm_page_t m;
+
+	pageoff = gpa & PAGE_MASK;
+	if (len > PAGE_SIZE - pageoff)
+		panic("vm_gpa_hold: invalid gpa/len: 0x%016lx/%lu", gpa, len);
+
+	count = 0;
+	for (i = 0; i < VM_MAX_MEMMAPS; i++) {
+		mm = &vm->mem_maps[i];
+		if (sysmem_mapping(vm, mm) && gpa >= mm->gpa &&
+		    gpa < mm->gpa + mm->len) {
+			vm_pointer_t gpap;
+
+			gpap = trunc_page(gpa);
+#ifdef __CHERI_PURE_CAPABILITY__
+			gpap = (vm_pointer_t)cheri_setaddress(vmm_gpa_root_cap,
+			    gpa);
+#endif
+			count = vm_fault_quick_hold_pages(&vm->vmspace->vm_map,
+			    (void * __capability)gpap, PAGE_SIZE, reqprot, &m,
+			    1);
+			break;
+		}
+	}
+
+	if (count == 1) {
+		*cookie = m;
+		return ((void *)(PHYS_TO_DMAP(VM_PAGE_TO_PHYS(m)) + pageoff));
+	} else {
+		*cookie = NULL;
+		return (NULL);
+	}
+}
+
+void *
+vm_gpa_hold(struct vcpu *vcpu, vm_paddr_t gpa, size_t len, int reqprot,
+	    void **cookie)
+{
+#ifdef INVARIANTS
+	/*
+	 * The current vcpu should be frozen to ensure 'vm_memmap[]'
+	 * stability.
+	 */
+	int state = vcpu_get_state(vcpu, NULL);
+	KASSERT(state == VCPU_FROZEN, ("%s: invalid vcpu state %d",
+	    __func__, state));
+#endif
+	return (_vm_gpa_hold(vcpu->vm, gpa, len, reqprot, cookie));
+}
+
+void *
+vm_gpa_hold_global(struct vm *vm, vm_paddr_t gpa, size_t len, int reqprot,
+    void **cookie)
+{
+	sx_assert(&vm->mem_segs_lock, SX_LOCKED);
+	return (_vm_gpa_hold(vm, gpa, len, reqprot, cookie));
+}
+
+void
+vm_gpa_release(void *cookie)
+{
+	vm_page_t m = cookie;
+
+	vm_page_unwire(m, PQ_ACTIVE);
+}
+
+int
+vm_get_register(struct vcpu *vcpu, int reg, uintcap_t *retval)
+{
+
+	if (reg >= VM_REG_LAST)
+		return (EINVAL);
+
+	return (vmmops_getreg(vcpu->cookie, reg, retval));
+}
+
+int
+vm_set_register(struct vcpu *vcpu, int reg, uintcap_t val)
+{
+	int error;
+
+	if (reg >= VM_REG_LAST)
+		return (EINVAL);
+	error = vmmops_setreg(vcpu->cookie, reg, val);
+	if (error || reg != VM_REG_GUEST_PC)
+		return (error);
+
+#if __has_feature(capabilities)
+#ifdef __CHERI_PURE_CAPABILITY__
+	vcpu->nextpc = (uintcap_t)cheri_setaddress(kernel_root_cap, val);
```

Comment:
This is used when userspace sets the initial PC value before executing guest instructions. In practice that's the only this is used, though in principle userspace bhyve could try to set it again. I agree that it's rather dodgy, but I'm not sure how else this could work.

I do need to reduce permissions on the cap in the purecap case.

Reply:
This probably needs to follow what gets done for ptrace and take in an untagged uintcap_t along with a separate tag, and then do a cheri_buildcap of that (and possible sealing for sentries) with the guest address space root cap to derive the unexposed tagged value

Reply:
I took a look at how ptrace works. I don't quite understand the corresponding problem for bhyve: what problem does it solve to require userspace bhyve to provide more than just an address? I'm still new to reasoning about the properties of sealed capabilities, so I'm sure I'm missing something, but I can't see how to resolve this.

Reply:
So I think what we might want to do is:
- Preserve the existing behaviour when the current value is unset, i.e., it is being initialized for the first time. In this case bhyve(8) is allowed to ask the kernel to create an almighty capability before the guest begins execution.
- Handle this the same way as ptrace_derive_cap() otherwise, i.e., try to derive a capability from an existing guest register.
- Add some mechanism similar to `cheri_ptrace_caps`, since the GDB stub could ideally be used to inject capabilities.